### ISSUE TITLE
CICERO-SCM

### ISSUE THREAD

**Issue by rgieseke** [openclimatedata/openscm#75]
https://brage.bibsys.no/xmlui/bitstream/handle/11250/192444/CICERO_Working_Paper_1999-02.pdf

**Comment by rgieseke** [comment]
Recent description of the SCM in the SI
https://iopscience.iop.org/1748-9326/12/2/024022/media/erlaa5b0a_suppdata.pdf
of this paper
https://iopscience.iop.org/article/10.1088/1748-9326/aa5b0a

**Comment by maritsandstad** [comment]
Hey, I am actually going to be (trying to) implement an openscm adaptor for the Cicero-SCM now. I guess what I should do is create my own branch for this and link it to this issue tracker. 

However, so far I haven't been able to run the vanilla tests yet. Do you have a document with your requirements, because I couldn't seem to find one...?

**Comment by rgieseke** [comment]
Hi @maritsandstad - that's great to hear! 

@swillner and @znicholls are still figuring out the first model integrations for e.g. Dice so the API definition is still under development, Cicero-SCM could become the next candidate!

Creating a fork and then later a Pull Request would be the appropriate process. What is the status of Cicero-SCM, can you run it from Python in general already?

The development process should be documented here:
https://openscm.readthedocs.io/en/latest/development.html

If something is missing or not working please open an issue :-)

All requirements and libraries are defined in the `setup.py` file.

If you're on Linux or macOS you can install a development environment with `make venv`

**Comment by znicholls** [comment]
Hi @maritsandstad,

As @rgieseke said, great news! I just wanted to reinforce one point. 

> the API definition is still under development,

The API is **very** likely to change so please be prepared to have to update things if required. This is all very new so there's lots of things we're struggling to work out and have to play around with a bit.

**Comment by maritsandstad** [comment]
Hi @rgieseke and @znicholls,

thanks for the fast replies. Currently it is not python operable, however, we were wanting to make the code more accessible (primarily in house, but also externally), and making it part of openscm seemed like a good an logical way of doing this. Hence, the fact that the API will change is not a huge problem for us. Having something to start with an updating from there is actually not a bad idea. Maybe we can even give input on things that would make the API more easily adaptible to  non-python models, as well?

Anyways, thanks so far, chances are I will have lots of questions later, however I'll try not to be too much of a nuisance...

cheers

**Comment by znicholls** [comment]
> Maybe we can even give input on things that would make the API more easily adaptible to non-python models, as well?

Would be super! 

> Anyways, thanks so far, chances are I will have lots of questions later, however I'll try not to be too much of a nuisance

Questions are good as it highlights things we need to address better in the docs (i.e. no matter how many you ask, you won't be a nuisance). Having said that, pull requests are even better as they can a) highlight the problem and b) address it at the same time :D

**Comment by maritsandstad** [comment]
Hi, so I tried to do the make venv, however, this doesn't really enforce requirements (for instance it goes ahead with whatever python version you have, even though the code actually requires 3.7) Perhaps however, you list a python 3.7 requirement somewhere that I haven't seen....

 The 3.7 is easily accommodated of course, but the pyam import of IamDataFrame is still giving me grief... It seems that there is some 'SymbolicTwoPointBVPLike' function that can't be found in the pycollocation package. Any ideas?

cheers,
marit

**Comment by rgieseke** [comment]
>  Perhaps however, you list a python 3.7 requirement somewhere that I haven't seen....

Good catch! We should make this explicit in the Makefile. (The required version was introduced due to using typing which wasn't available in earlier versions i think.)

**Comment by rgieseke** [comment]
> The 3.7 is easily accommodated of course, but the pyam import of IamDataFrame is still giving me grief... It seems that there is some 'SymbolicTwoPointBVPLike' function that can't be found in the pycollocation package. Any ideas?

That's most likely the wrong `pyam` library, you need https://github.com/IAMconsortium/pyam available via `pip install pyam-iamc`

**Comment by maritsandstad** [comment]
Thank you, that did the trick!

**Comment by maritsandstad** [comment]
Did I understand correctly; you want a fork and then pull request, not a branch and then pull request? Does this mean that any implementation of an adapter for a new SCM is meant to by a separate project not to be merged into the mainline? (Probably stupid question, but I'm not really that well-versed in git yet... )

**Comment by rgieseke** [comment]
Sorry, if i was unclear, or i misunderstood the question - the usual GitHub workflow for external contributors is to fork and then create a Pull Request into the repo. 

As for inclusion of code wrapping/interfacing Cicero-SCM in OpenSCM, i think both ways are possible, either include all wrapping code in OpenSCM or have a separate module which uses OpenSCM for its API. Since OpenSCM is still under heavy development we haven't fully figured this out yet. Really simple models like DICE are easy to include, but if external dependencies need to included than it becomes more difficult.

For example Pymagic ships with a bundled binary of MAGICC (since it is not open source yet), Pyhector requires compilation of the C++ libraries of Hector etc.

It also depends on your Open-Source strategy, do you plan to release Cicero-SCM as open source?

**Comment by maritsandstad** [comment]
Yes, I think so, but it might be a multistep process. Right now it is in fortran, and gives nice results, but might have some of the impairments to readability that you typically see in fortran codes (written by physicists...)

Thanks for the answer, though, I'll make me a fork, then;-)

**Comment by maritsandstad** [comment]
For reference, the fork is here: https://github.com/ciceroOslo/openscm

**Comment by maritsandstad** [comment]
Hey, so lots of other things to do, but my fork is up, and I've been following the https://openscm.readthedocs.io/en/latest/development.html strategy to get my adapter started, adding only passes in all the method for now. However, when I run my pytest now, I get lots of errors due to a lack of the test_adapter fixture. I've been googling a bit, but still don't really understand how to handle the issue... Any tips?

**Comment by znicholls** [comment]
Hey Marit,

This is all pretty all over the place on our side (still lots of moving
parts at the moment, see the pull requests) so I can understand why it
might be super confusing! The easiest thing would be to a) hold off on your
adapter for ~2 weeks, by which time we’ll have much clearer docs about
writing adapters or b) focus on how to run ciceroscm from python. Do you
have a pyciceroscm package or similar? If not, perhaps @rgieseke and
@swillner can help as they’re the experts on writing these python wrappers.
Failing that, I’ll try and give a better explanation in the morning (bed
time in Australia) if no one else has.

Cheers,

Zeb

**Comment by maritsandstad** [comment]
Nevermind, updated my fork and got the tests to work. However, you changed some things (location of the abstract Adapter class and a new abstract method in the _Adaptertester) which both warrant changes to the developement docs: https://openscm.readthedocs.io/en/latest/development.html

**Comment by maritsandstad** [comment]
Thanks for your answer, I'll try and see what I can do to get a python package up and stay aware of the moving parts. Good night then, I guess

**Comment by znicholls** [comment]
> However, you changed some things (location of the abstract Adapter class and a new abstract method in the _Adaptertester) which both warrant changes to the developement docs

Great thanks, I've made an issue (#188) to remind myself to do this as part of adding the PH99 adapter (#60)

**Comment by maritsandstad** [comment]
Just to check, would about now be a good time to sync and start making the adapter?

**Comment by znicholls** [comment]
I would wait until the current pull requests go through. Once they're through, I'll start writing proper docs about how to write an adapter which should hopefully make your life much easier.

**Comment by maritsandstad** [comment]
Great thanks for your answer, I'm all about the easy life;-)
